src/lib.rs

```diff
@@ -4,21 +4,136 @@
+//! different consumers across several different domains, and is used by both
+//! Hubris and Humility.
+//!
+//! Our nomenclature:
+//!
+//! - *Dump area*:  a contiguous area of RAM within Hubris that holds all or
+//!   part of a dump.  This is an area of RAM that is not otherwise used by
+//!   the system and (obviously?) shouldn't itself be deumped.
+//!
+//! - *Dump contents*:  the contents of a dump area, and can be either a task
+//!   (or part of a task) or the entire system (or part of it).  If an area is
+//!   used for part of the system, all dump areas are used to dump the system.
+//!   (That is, task dumps cannot be interspersed with system dumps.)
+//!
+//! - *Dump segment header*:  a header describing a contiguous region of memory
+//!   to be written into a dump.  These are added to a dump area via
+//!   [`add_dump_segment_header`].
+//!
+//! - *Dump segment*:  the actual data itself in a dump.  This can have the form
+//!   of data (always compressed), or some limited metadata (registers and
+//!   task information, if any).  If there is both metadata and data, the
+//!   metadata will always precede the data.
+//!
+//! - *Dump agent proxy*.  The body of software that creates dump areas and
+//!   doles them out to dump agents.
+//!
+//! - *Dump agent*. The software that claims a dump area from the dump agent
+//!   proxy for the purpose of arranging dumping into it.
+//!
+//! - *Dumper*.  The body of software that actually performs the dumping:
+//!   knowing only the address of a dump area, it will dump contents into dump
+//!   areas such as they are available.  (Jefe, the dedicated dumper task, and
+//!   Humility in its emulation modes can all act as the dumper.)
+//!
+//! In Hubris, Jefe acts as the dump agent proxy, and (in the case of task
+//! dumps) also serves as the dump agent and the dumper.  For system dumps,
+//! the dedicated dump agent serves as the agent, and an outside system
+//! (either Humility in its emulation modes or a disjoint microcontroller
+//! connected via SWD) acts as the dumper.
+//!
+//! Regardless of which bodies are playing which part, the flow is:
+//!
+//!  1. Dump agent asks dump agent proxy to claim an area on its behalf.
+//!
+//!  2. Dump agent adds segment headers to describe the data to be dumped.
+//!
+//!  3. Dumper calls [`dump`] to actually do the dumping.  [`dump`] will
+//!     read and write to memory via the passed closures.
+//!
+//!  4. Dump is retrieved by Humility for decompressing and processing.
+//!
+
+use hubpack::SerializedSize;
+use serde::{Deserialize, Serialize};
 use zerocopy::{AsBytes, FromBytes};
 
 pub const DUMP_MAGIC: [u8; 4] = [0x1, 0xde, 0xde, 0xad];
 pub const DUMP_UNINITIALIZED: [u8; 4] = [0xba, 0xd, 0xca, 0xfe];
 pub const DUMP_SEGMENT_PAD: u8 = 0x55;
+
+const DUMP_SEGMENT_ALIGN: usize = 4;
+const DUMP_SEGMENT_MASK: usize = DUMP_SEGMENT_ALIGN - 1;
+
 pub const DUMP_REGISTER_MAGIC: [u8; 2] = [0xab, 0xba];
```

Comment:
Why are these `[u8; 2]` instead of `u16`?

(If they need to be unaligned, they could be `zerocopy::byteorder::U16<zerocopy::byteorder::LE>`)

Reply:
For magic values, I have always had a personal preference to byte arrays vs. integer quantities because it takes factors out endianness. Which is a long way of saying: no good reason, really -- just personal preference. ;)

Reply:
Okay, that's fine by me